apps/app_manager.py

```diff
@@ -51,99 +54,115 @@ def load_all_apps(self):
-        #If user runs in single-app mode and by accident 
-        #autocompletes the app name too far, it shouldn't fail
-        main_py_string = ".main.py" 
+        # If user runs in single-app mode and by accident
+        # autocompletes the app name too far, it shouldn't fail
+        main_py_string = ".main.py"
         if app_import_path.endswith(main_py_string):
             app_import_path = app_import_path[:-len(main_py_string)]
-        app = importlib.import_module(app_import_path+'.main', package='apps')
-        app.init_app(self.i, self.o)
+        app = importlib.import_module(app_import_path + '.main', package='apps')
+        if is_class_based_module(app):
+            zero_app_subclass = get_zeroapp_class_in_module(app)
+            app = zero_app_subclass(self.i, self.o)
+        else:
+            app.init_app(self.i, self.o)
         return app
 
     def get_subdir_menu_name(self, subdir_path):
         """This function gets a subdirectory path and imports __init__.py from it. It then gets _menu_name attribute from __init__.py and returns it. 
         If failed to either import __init__.py or get the _menu_name attribute, it returns the subdirectory name."""
         subdir_import_path = subdir_path.replace('/', '.')
         try:
-            subdir_object = importlib.import_module(subdir_import_path+'.__init__')
+            subdir_object = importlib.import_module(subdir_import_path + '.__init__')
             return subdir_object._menu_name
         except Exception as e:
             print("Exception while loading __init__.py for subdir {}".format(subdir_path))
             print(e)
             return os.path.split(subdir_path)[1].capitalize()
 
-    def get_ordering(self, path, cache={}):
+    def get_ordering(self, path, cache=None):
         """This function gets a subdirectory path and imports __init__.py from it. It then gets _ordering attribute from __init__.py and returns it. It also caches the attribute for faster initialization.
         If failed to either import __init__.py or get the _ordering attribute, it returns an empty list."""
+        if cache is None:
+            cache = {}
         if path in cache:
             return cache[path]
         import_path = path.replace('/', '.')
+        ordering = []
         try:
-            imported_module = importlib.import_module(import_path+'.__init__')
+            imported_module = importlib.import_module(import_path + '.__init__')
             ordering = imported_module._ordering
         except ImportError as e:
             print("Exception while loading __init__.py for directory {}".format(path))
             print(e)
-            ordering = []
-        except AttributeError as e:
-            #print("No ordering for directory {}".format(path))
-            #print(e)
-            ordering = []
+        except AttributeError as e:  # todo: refactor as a single 'except' statement
+            print("No ordering for directory {}".format(path))
```

Comment:
I noticed that this warning triggers basically every time. The feature doesn't seem used that much.

Reply:
Yep, that's a warning left over from debugging. I'll do a proposal about these in a couple minutes =)

Reply:
So, I don't think this warning is necessary at all, considering that the feature is hardly used (except for main menu, where it's absolutely crucial to have this feature). I'd 1) put a `pass` in that AttributeError block 2) put a `logger.debug("Found ordering for {} directory!".format(import_path))`, so that we know when ordering is found when we need to debug this.

Reply:
Also, are you sure you need a single 'except' statement? They're handling two different cases. If you want, you can avoid processing `AttributeError` by doing a `hasattr(imported_module, "ordering")` check.

Reply:
With your proposal we treat both cases differently, so two except statements are better indeed.